File: tools/authenticode-digest-reference.py
#!/usr/bin/env python3
"""Reference implementation of the Authenticode PE digest.

A deliberately naive, dependency-free transcription of the hashing algorithm in
"Windows Authenticode Portable Executable Signature Format", kept separate from the
C# engine so the two can disagree. It backs the known-answer vectors in
PeDigestTests and doubles as an offline oracle on machines without signtool.

Usage:
  authenticode-digest-reference.py FILE...            print each file's digest
  authenticode-digest-reference.py --check DIR...     self-check against signed PEs

--check walks the directories for Authenticode-signed PE images, recomputes each
digest, and confirms it appears inside that file's own embedded signature. Any
directory of Microsoft-signed binaries works; a NuGet package cache is a convenient
source of several thousand, covering both PE32 and PE32+.
"""

import hashlib
import os
import struct
import sys

CERTIFICATE_DIRECTORY_INDEX = 4


def parse(image):
    """Return the header offsets the digest depends on."""
    if image[:2] != b"MZ":
        raise ValueError("not a PE image: missing 'MZ' signature")

    pe = struct.unpack_from("<I", image, 0x3C)[0]
    if image[pe:pe + 4] != b"PE\0\0":
        raise ValueError("not a PE image: missing 'PE\\0\\0' signature")

    section_count = struct.unpack_from("<H", image, pe + 6)[0]
    optional_size = struct.unpack_from("<H", image, pe + 20)[0]
    optional = pe + 24

    magic = struct.unpack_from("<H", image, optional)[0]
    if magic not in (0x10B, 0x20B):
        raise ValueError("unsupported optional header magic 0x%X" % magic)
    pe32_plus = magic == 0x20B

    # SizeOfHeaders and CheckSum sit at the same optional-header offsets in both
    # formats; the data directories do not.
    directories = optional + (112 if pe32_plus else 96)
    certificate_directory = directories + CERTIFICATE_DIRECTORY_INDEX * 8
    rva_count = struct.unpack_from("<I", image, optional + (108 if pe32_plus else 92))[0]
    if rva_count <= CERTIFICATE_DIRECTORY_INDEX:
        raise ValueError("image has no attribute-certificate data directory")

    table_offset, table_size = struct.unpack_from("<II", image, certificate_directory)

    sections = []
    section_table = optional + optional_size
    for i in range(section_count):
        entry = section_table + i * 40
        raw_size = struct.unpack_from("<I", image, entry + 16)[0]
        raw_offset = struct.unpack_from("<I", image, entry + 20)[0]
        if raw_size:
            sections.append((raw_offset, raw_size))

    return {
        "pe32_plus": pe32_plus,
        "checksum": optional + 64,
        "certificate_directory": certificate_directory,
        "table_offset": table_offset,
        "table_size": table_size,
        "size_of_headers": struct.unpack_from("<I", image, optional + 60)[0],
        "sections": sections,
    }


def digest(image, headers, algorithm="sha256"):
    """Compute the Authenticode digest of an in-memory PE image."""
    accumulator = hashlib.new(algorithm)

    # Headers, minus the CheckSum field and the certificate directory entry: the two
    # ranges signing itself rewrites.
    accumulator.update(image[0:headers["checksum"]])
    accumulator.update(image[headers["checksum"] + 4:headers["certificate_directory"]])
    accumulator.update(image[headers["certificate_directory"] + 8:headers["size_of_headers"]])

    hashed = headers["size_of_headers"]
    for offset, size in sorted(headers["sections"]):
        accumulator.update(image[offset:offset + size])
        hashed += size

    # Whatever trails the last section, except the certificate table itself.
    trailing = len(image) - headers["table_size"] - hashed
    if trailing > 0:
        accumulator.update(image[hashed:hashed + trailing])

    return accumulator.hexdigest()


def check(paths):
    """Recompute digests of signed PEs and look for each inside its own signature."""
    matched = mismatched = 0
    for root in paths:
        for directory, _, names in os.walk(root):
            for name in names:
                path = os.path.join(directory, name)
                try:
                    with open(path, "rb") as handle:
                        image = handle.read()
                    headers = parse(image)
                except (ValueError, OSError, struct.error):
                    continue
                if not headers["table_size"]:
                    continue

                table = image[headers["table_offset"]:headers["table_offset"] + headers["table_size"]]
                for algorithm in ("sha256", "sha384", "sha512", "sha1"):
                    if bytes.fromhex(digest(image, headers, algorithm)) in table:
                        matched += 1
                        break
                else:
                    mismatched += 1
                    print("MISMATCH %s" % path)

    print("%d signed images matched, %d mismatched" % (matched, mismatched))
    return 1 if mismatched else 0


def main(argv):
    if len(argv) < 2:
        print(__doc__.strip())
        return 2

    if argv[1] == "--check":
        return check(argv[2:] or ["."])

    for path in argv[1:]:
        with open(path, "rb") as handle:
            image = handle.read()
        print("%s  %s" % (digest(image, parse(image)), path))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: test/SignUniversal.Core.Tests/AuthenticodeDigestReference.cs
using System.Buffers.Binary;

namespace SignUniversal.Core.Tests;

/// <summary>
/// A second, deliberately naive implementation of the Authenticode digest, transcribed
/// straight from the 15-step procedure in "Windows Authenticode Portable Executable
/// Signature Format".
/// </summary>
/// <remarks>
/// <para>
/// It shares no code with <see cref="SignUniversal.Core.Authenticode.PeFile"/> and is
/// written for obviousness rather than efficiency: whole file in memory, no streaming, no
/// reuse. That is the point — it exists to disagree with the engine when the engine is
/// wrong, and the engine's streaming walk is where a transcription slip would hide.
/// </para>
/// <para>
/// Its own credibility comes from the corpus check in <c>PeDigestTests</c>: pointed at a
/// directory of already-signed binaries, it recomputes each digest and finds it inside
/// that file's own signature. That sweep is what originally settled the questions the
/// specification leaves ambiguous.
/// </para>
/// </remarks>
internal static class AuthenticodeDigestReference
{
    /// <summary>Computes the Authenticode digest of an in-memory PE image.</summary>
    /// <param name="image">The whole file.</param>
    /// <param name="hashAlgorithm">The digest algorithm.</param>
    /// <returns>The digest.</returns>
    public static byte[] ComputeDigest(byte[] image, HashAlgorithmName hashAlgorithm)
    {
        PeLayout layout = PeLayout.Read(image);

        using IncrementalHash hash = IncrementalHash.CreateHash(hashAlgorithm);

        // Headers, minus the checksum field and the certificate directory entry.
        hash.AppendData(image, 0, layout.ChecksumOffset);
        hash.AppendData(
            image,
            layout.ChecksumOffset + 4,
            layout.CertificateDirectoryOffset - (layout.ChecksumOffset + 4));
        hash.AppendData(
            image,
            layout.CertificateDirectoryOffset + 8,
            layout.SizeOfHeaders - (layout.CertificateDirectoryOffset + 8));

        int hashed = layout.SizeOfHeaders;

        foreach ((int offset, int size) in layout.Sections.OrderBy(section => section.Offset))
        {
            hash.AppendData(image, offset, size);
            hashed += size;
        }

        // Trailing data, except the certificate table itself.
        int trailing = image.Length - layout.CertificateTableSize - hashed;
        if (trailing > 0)
        {
            hash.AppendData(image, hashed, trailing);
        }

        return hash.GetHashAndReset();
    }

    /// <summary>Reads the attribute certificate table of a signed image.</summary>
    /// <param name="image">The whole file.</param>
    /// <returns>The table bytes, or <see langword="null"/> when the image is unsigned.</returns>
    public static byte[]? TryReadCertificateTable(byte[] image)
    {
        PeLayout layout = PeLayout.Read(image);

        if (layout.CertificateTableSize == 0)
        {
            return null;
        }

        return image.AsSpan(layout.CertificateTableOffset, layout.CertificateTableSize).ToArray();
    }

    private readonly record struct PeLayout(
        int ChecksumOffset,
        int CertificateDirectoryOffset,
        int CertificateTableOffset,
        int CertificateTableSize,
        int SizeOfHeaders,
        IReadOnlyList<(int Offset, int Size)> Sections)
    {
        public static PeLayout Read(byte[] image)
        {
            if (image.Length < 0x40 || BinaryPrimitives.ReadUInt16LittleEndian(image) != 0x5A4D)
            {
                throw new InvalidDataException("not a PE image");
            }

            int pe = (int)BinaryPrimitives.ReadUInt32LittleEndian(image.AsSpan(0x3C));
            if (BinaryPrimitives.ReadUInt32LittleEndian(image.AsSpan(pe)) != 0x00004550)
            {
                throw new InvalidDataException("not a PE image");
            }

            int sectionCount = BinaryPrimitives.ReadUInt16LittleEndian(image.AsSpan(pe + 6));
            int optionalHeaderSize = BinaryPrimitives.ReadUInt16LittleEndian(image.AsSpan(pe + 20));
            int optional = pe + 24;

            int magic = BinaryPrimitives.ReadUInt16LittleEndian(image.AsSpan(optional));
            bool pe32Plus = magic switch
            {
                0x10B => false,
                0x20B => true,
                _ => throw new InvalidDataException($"unsupported optional header magic 0x{magic:X}"),
            };

            int directories = optional + (pe32Plus ? 112 : 96);
            int rvaCount = (int)BinaryPrimitives.ReadUInt32LittleEndian(
                image.AsSpan(optional + (pe32Plus ? 108 : 92)));

            if (rvaCount <= 4)
            {
                throw new InvalidDataException("no attribute-certificate data directory");
            }

            int certificateDirectory = directories + (4 * 8);
            List<(int Offset, int Size)> sections = [];
            int sectionTable = optional + optionalHeaderSize;

            for (int i = 0; i < sectionCount; i++)
            {
                int entry = sectionTable + (i * 40);
                int rawSize = (int)BinaryPrimitives.ReadUInt32LittleEndian(image.AsSpan(entry + 16));
                int rawOffset = (int)BinaryPrimitives.ReadUInt32LittleEndian(image.AsSpan(entry + 20));

                if (rawSize > 0)
                {
                    sections.Add((rawOffset, rawSize));
                }
            }

            return new PeLayout(
                optional + 64,
                certificateDirectory,
                (int)BinaryPrimitives.ReadUInt32LittleEndian(image.AsSpan(certificateDirectory)),
                (int)BinaryPrimitives.ReadUInt32LittleEndian(image.AsSpan(certificateDirectory + 4)),
                (int)BinaryPrimitives.ReadUInt32LittleEndian(image.AsSpan(optional + 60)),
                sections);
        }
    }
}
